packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -6688,31 +6688,42 @@ abstract class RenderObjectElement extends Element {
   }
 
   void _updateParentData(ParentDataWidget<ParentData> parentDataWidget) {
-    bool applyParentData = true;
-    assert(() {
-      try {
-        if (!parentDataWidget.debugIsValidRenderObject(renderObject)) {
-          applyParentData = false;
+    if (parentDataWidget.debugIsValidRenderObject(renderObject)) {
+      parentDataWidget.applyParentData(renderObject);
+    } else {
+      // If it's not a valid render object, then don't show the ErrorWidget, or
+      // throw an exception, just report the error and move on.
+      //
+      // This prevents it from crashing in release mode and still reports the
+      // original error in debug mode.
+      assert(() {
+        try {
+          final String parentDataWidgetType = parentDataWidget.toStringShort();
+          String errorMessage = 'Incorrect use of ParentDataWidget.\n';
+
+          if (parentDataWidgetType == 'Positioned') {
+            errorMessage += 'The widget `Positioned` must be a descendant of a `Stack` widget.\n';
+          } else if (parentDataWidgetType == 'Flexible' || parentDataWidgetType == 'Expanded') {
+            errorMessage += 'The widget `$parentDataWidgetType` must be a direct child of a `Row`, `Column`, or `Flex` widget.\n';
+          }
+
```

Comment:
This checking for types is not something we should do. There are many ParentDataWidgets, this error message handling could not possibly account for all of them. It also means we are baking in knowledge of other classes here.

Reply:
Sure, the idea here is to only account for 3 common ones -- it leaves the error message as-is for any other ParentDataWidget to reduce confusion. IMO "Incorrect use of ParentDataWidget" isn't a very clear error message.

Having said this I'm very happy to remove attempted improvements to DX and fix only the original issue!

Reply:
Ah I see, makes sense. I think that is beyond the scope of this fix then, would you agree? I will remove changes to the error messages in this PR tonight. 

A problem I'm having with this fix is testing. At its core, it's fixing some bad code that is only run in debug mode. I left a comment [here](https://github.com/flutter/flutter/pull/155091#discussion_r1766971811), do you know of a way to test this?